### ISSUE TITLE
Action's Effect.asyncReply obfuscates exceptions 

### ISSUE THREAD

**Issue by octonato** [lightbend/kalix-jvm-sdk#954]
I've seen the following situation when implementing the samples for [reliable timers](https://github.com/lightbend/kalix-jvm-sdk/pull/921). 

Given an entity with carefully crafted exceptions:

```scala
  // Method in Value Entity returning NOT_FOUND
  override def confirm(currentState: OrderState, confirmRequest: example.ConfirmRequest): ValueEntity.Effect[Empty] =
    if (currentState.placed) // <2>
      effects
        .updateState(currentState.copy(confirmed = true))
        .thenReply(Empty.defaultInstance)
    else
      effects
        .error(
          s"No order found for ${confirmRequest.orderNumber}", 
          Status.Code.NOT_FOUND) // <3>
```

When calling it directly with...

```scala
curl -XPOST -v -H "Content-Type: application/json" http://localhost:9000/com.example.OrderService/Confirm -d '{ "order_number": "non-existent-order-number" }'
```
we get ...

```scala
No order found for 'non-existent-order-number'
```

Then, with an Actions calling that same method...

```scala
// now calling from Action using asyncReply
override def confirm(orderNumber: OrderNumber): Action.Effect[Empty] = {
    logger.info("Confirming order {}", orderNumber.number)
    val reply =
      for {
        _ <- components.order // <1>
          .confirm(example.ConfirmRequest(orderNumber.number))
          .execute()
        _ <- timers.cancel(timerName(orderNumber)) // <2>
      } yield Empty.defaultInstance

    effects.asyncReply(reply)
  }
```
we get...

```scala
Unexpected error [64017428-f4b8-45a9-b228-882770f01a22]
```

**Comment by octonato** [comment]
The cause for this may be #955. 

If the status code is not being propagated properly, then I can imagine that `asyncReply` will hide the error message because it gets a 500 from the entity.

**Comment by johanandren** [comment]
That's a regular gRPC call between the scene, the error coming back in an async call will be a grpc status exception. We treat thrown exceptions, both in async and sync code, as potentially leaking internal details of the service, so that's why we give it a correlation id to hand back to the user.


I think automagically passing such error codes through from a called service is the wrong thing to do. Another component failing doesn't inherently translate 1:1 to the same error code from the calling component back to the client/user.

A user would have to catch/recover the failed future and turn the exception into a suitable error response, which may or may not have the same error code that it got from the other service.

**Comment by octonato** [comment]
A user defines an error message in the entity. If we call that entity directly using a gRPC client, we do get the message the user defined. So, interacting with it directly yields what what the user intended. 

But if we put an Action in between, the Action drops the message defined by the user and make it an "Unexpected error". 

We tell the users that using `effects().error()` is what they should use to report users to clients. When we promote the usage of Action as controllers, it's kind of surprising to see it interfering with the error signaling. 

I understand that if I call a component and `flatMap` with some other method, and that other method fails, I do want to convert to a "Unexpected error`, but as long the failure we have at hand is one explicitly defined by the user, I think we should propagate and do what the user requested.

In #956, I was mentioning that we should not have `io.grpc` in our APIs. If we use our own exception hierarchy whenever using `effects().error()`, we could have more control over it. And it will fit better with a code-first approach.

**Comment by johanandren** [comment]
I still disagree, because the error returned by the other component is decided upon in isolation, say NotFound/404 but that doesn't mean it by definition is a good error code and message for the aggregate action failing, where it is just one of the (internal) tasks it expected to do.

I think we should document that cross-component calls can throw gRPC errors and have a sample in the aggregation docs of how to catch and change them into an error from the aggregating action (including passing them through for cases where that makes sense for the specific composed action).

**Comment by octonato** [comment]
But I don't think they are decided in isolation. They make part of the same service. The fact that I can call the entity directly is because we don't have yet the ACL. 

If a user can't define their error codes at the entity level and let then propagate to the action, they will have to catch and emit them again using action's `effect.error` (by copying the codes). That's the equivalent of catch and re-throwing without any added value for the user. It's just an annoyance. 

I don't think this is a good DX. We are just adding the burden to the developers to do nonsensical and repetitive work.

**Comment by johanandren** [comment]
No, if the developer of the service did not expect the called service to return an error, that is an unhandled internal error of the component, and is therefore logged for the owner of the service to debug/fix/handle/expect, but the client calling the service cannot be expected to do anything about that, so should get a 500 internal error (with the correlation id so they can tell the owner of the service who can fix it).

The alternative is like propagating a connection to db failed back to the client in a classical 3-tier architecture, which is close to useless and leaks internal details of the service.
